rfcs/59-from_reflect_ergonomics.md

```diff
@@ -0,0 +1,284 @@
+# Feature Name: `from_reflect_ergonomics`
+
+## Summary
+
+Derive `FromReflect` by default when deriving `Reflect`, add `ReflectFromReflect` type data, and make `ReflectDeserializer` easier to use.
+
+## Terminology
+
+#### *Dynamic Types*
+
+To make this document easier to read, the term "**Dynamic**" (with this casing) will be used to refer to any of the `Dynamic***` structs. This includes `DynamicStruct`, `DynamicTupleStruct`, `DynamicList`, etc.
+
+Other forms might include "Dynamics," "Dynamic structs," or "Dynamic types" depending on context.
+
+In example code, variables containing these types will be denoted as `dyn_X`, where `X` corresponds to their data structure. For example, `dyn_tuple_struct` would represent a `DynamicTupleStruct`.
+
+#### *Real Types*
+
+The term "**real type**" will be used to distinguish Dynamics and other concrete types. We can define it as any concrete type known at compile time that is *not* one of the Dynamic structs. 
+
+In example code, variables containing real types will be denoted as `real_Y`, where `Y` corresponds to their data structure. For example, `real_list` would represent a list-like type, such as `Vec<T>`.
+
+## Background
+
+### What is `FromReflect`?
+
+For those unfamiliar, here's a quick refresher on `FromReflect`.
+
+The `FromReflect` trait is meant to allow types to be constructed dynamically (well, not completely, but we'll get into that later). Specifically, it enables a real type to be generated from a Dynamic one. As an example, we can convert a `DynamicTuple` to a real tuple like so:
+
+```rust
+let mut dyn_tuple = DynamicTuple::default();
+dyn_tuple.insert(123usize);
+dyn_tuple.insert(321usize);
+
+let real_tuple = <(usize, usize)>::from_reflect(&dyn_tuple).unwrap();
+```
+
+A Dynamic value to a real one. Neat!
+
+### How does `FromReflect` work?
+
+Basically it's `FromReflect` all the way down. 
+
+Given a struct like:
+
+```rust
+struct Foo {
+  value: (f32, f32)
+}
+```
+
+Calling `Foo::from_reflect` will kick off a chain of `FromReflect`:
+
+`Foo::from_reflect` calls `<(f32, f32)>::from_reflect` which itself calls `f32::from_reflect`.
+
+This is why, when deriving `FromReflect`, all fields *must* also implement `FromReflect`[^1]. Without this, the trait itself won't just "not work"— it won't even compile.
```

Comment:
To be fair, "this invariant is validated at compile time" is not much simpler haha.

I can reduce the wordiness, but I'm not sure about using language like "invariant" as I think it could be confusing to newer/intermediate devs (including myself who still struggles with understanding variance lol 😅). 

Thoughts on this?

Reply:
I'm in favor of just cutting it down to "Without this, the trait implementation won't compile". 

Irrelevant to this PR, but I'm considering if it would be a good idea to rename `ArchetypeInvariants` to `ArchetypeRequirements` for the same reason. I feel like a newbie would prefer to see "{Jargon}Requirements" and get at least some idea as to what it means rather than "{Jargon}{Jargonier}" and just ignore it entirely.